// @metamask/utils/dist/node.mjs
import {
  createSandbox,
  directoryExists,
  ensureDirectoryStructureExists,
  fileExists,
  forceRemove,
  readFile,
  readJsonFile,
  writeFile,
  writeJsonFile
} from "./chunk-52OU772R.mjs";
import "./chunk-622IOGVI.mjs";
import {
  bigIntToHex,
  hexToBigInt,
  hexToNumber,
  numberToHex
} from "./chunk-S3UHBN2Z.mjs";
import "./chunk-AY6FDCBT.mjs";
import {
  createDeferredPromise
} from "./chunk-B7LIM2PK.mjs";
import {
  Duration,
  inMilliseconds,
  timeSince
} from "./chunk-THNLGEXV.mjs";
import "./chunk-RRYOWRCV.mjs";
import {
  VersionRangeStruct,
  VersionStruct,
  assertIsSemVerRange,
  assertIsSemVerVersion,
  gtRange,
  gtVersion,
  isValidSemVerRange,
  isValidSemVerVersion,
  satisfiesVersionRange
} from "./chunk-ROQSKDP5.mjs";
import {
  JsonRpcErrorStruct,
  JsonRpcFailureStruct,
  JsonRpcIdStruct,
  JsonRpcNotificationStruct,
  JsonRpcParamsStruct,
  JsonRpcRequestStruct,
  JsonRpcResponseStruct,
  JsonRpcSuccessStruct,
  JsonRpcVersionStruct,
  JsonStruct,
  PendingJsonRpcResponseStruct,
  UnsafeJsonStruct,
  assertIsJsonRpcError,
  assertIsJsonRpcFailure,
  assertIsJsonRpcNotification,
  assertIsJsonRpcRequest,
  assertIsJsonRpcResponse,
  assertIsJsonRpcSuccess,
  assertIsPendingJsonRpcResponse,
  exactOptional,
  getJsonRpcIdValidator,
  getJsonSize,
  getSafeJson,
  isJsonRpcError,
  isJsonRpcFailure,
  isJsonRpcNotification,
  isJsonRpcRequest,
  isJsonRpcResponse,
  isJsonRpcSuccess,
  isPendingJsonRpcResponse,
  isValidJson,
  jsonrpc2,
  object
} from "./chunk-6C35XQOF.mjs";
import "./chunk-I575FZFH.mjs";
import {
  createModuleLogger,
  createProjectLogger
} from "./chunk-RIRDOQPX.mjs";
import {
  CAIP_ACCOUNT_ADDRESS_REGEX,
  CAIP_ACCOUNT_ID_REGEX,
  CAIP_CHAIN_ID_REGEX,
  CAIP_NAMESPACE_REGEX,
  CAIP_REFERENCE_REGEX,
  CaipAccountAddressStruct,
  CaipAccountIdStruct,
  CaipChainIdStruct,
  CaipNamespaceStruct,
  CaipReferenceStruct,
  isCaipAccountAddress,
  isCaipAccountId,
  isCaipChainId,
  isCaipNamespace,
  isCaipReference,
  parseCaipAccountId,
  parseCaipChainId
} from "./chunk-TGOMTREC.mjs";
import {
  ChecksumStruct
} from "./chunk-GZS3IQBZ.mjs";
import {
  base64
} from "./chunk-NQMRFZHB.mjs";
import {
  createBigInt,
  createBytes,
  createHex,
  createNumber
} from "./chunk-O3EIM33O.mjs";
import {
  HexAddressStruct,
  HexChecksumAddressStruct,
  HexStruct,
  StrictHexStruct,
  add0x,
  assertIsBytes,
  assertIsHexString,
  assertIsStrictHexString,
  base64ToBytes,
  bigIntToBytes,
  bytesToBase64,
  bytesToBigInt,
  bytesToHex,
  bytesToNumber,
  bytesToSignedBigInt,
  bytesToString,
  concatBytes,
  createDataView,
  getChecksumAddress,
  hexToBytes,
  isBytes,
  isHexString,
  isStrictHexString,
  isValidChecksumAddress,
  isValidHexAddress,
  numberToBytes,
  remove0x,
  signedBigIntToBytes,
  stringToBytes,
  valueToBytes
} from "./chunk-VA2DRBDE.mjs";
import {
  AssertionError,
  assert,
  assertExhaustive,
  assertStruct
} from "./chunk-74DGVJVE.mjs";
import {
  getErrorMessage,
  isErrorWithCode,
  isErrorWithMessage,
  isErrorWithStack,
  wrapError
} from "./chunk-XYGUOY6N.mjs";
import {
  ESCAPE_CHARACTERS_REGEXP,
  JsonSize,
  calculateNumberSize,
  calculateStringSize,
  getKnownPropertyNames,
  hasProperty,
  isASCII,
  isNonEmptyArray,
  isNullOrUndefined,
  isObject,
  isPlainObject
} from "./chunk-H4YFDLB7.mjs";
import {
  FrozenMap,
  FrozenSet
} from "./chunk-JPAL7Q5S.mjs";
import "./chunk-X66SUIEF.mjs";
import "./chunk-WA4DHIND.mjs";
export {
  AssertionError,
  CAIP_ACCOUNT_ADDRESS_REGEX,
  CAIP_ACCOUNT_ID_REGEX,
  CAIP_CHAIN_ID_REGEX,
  CAIP_NAMESPACE_REGEX,
  CAIP_REFERENCE_REGEX,
  CaipAccountAddressStruct,
  CaipAccountIdStruct,
  CaipChainIdStruct,
  CaipNamespaceStruct,
  CaipReferenceStruct,
  ChecksumStruct,
  Duration,
  ESCAPE_CHARACTERS_REGEXP,
  FrozenMap,
  FrozenSet,
  HexAddressStruct,
  HexChecksumAddressStruct,
  HexStruct,
  JsonRpcErrorStruct,
  JsonRpcFailureStruct,
  JsonRpcIdStruct,
  JsonRpcNotificationStruct,
  JsonRpcParamsStruct,
  JsonRpcRequestStruct,
  JsonRpcResponseStruct,
  JsonRpcSuccessStruct,
  JsonRpcVersionStruct,
  JsonSize,
  JsonStruct,
  PendingJsonRpcResponseStruct,
  StrictHexStruct,
  UnsafeJsonStruct,
  VersionRangeStruct,
  VersionStruct,
  add0x,
  assert,
  assertExhaustive,
  assertIsBytes,
  assertIsHexString,
  assertIsJsonRpcError,
  assertIsJsonRpcFailure,
  assertIsJsonRpcNotification,
  assertIsJsonRpcRequest,
  assertIsJsonRpcResponse,
  assertIsJsonRpcSuccess,
  assertIsPendingJsonRpcResponse,
  assertIsSemVerRange,
  assertIsSemVerVersion,
  assertIsStrictHexString,
  assertStruct,
  base64,
  base64ToBytes,
  bigIntToBytes,
  bigIntToHex,
  bytesToBase64,
  bytesToBigInt,
  bytesToHex,
  bytesToNumber,
  bytesToSignedBigInt,
  bytesToString,
  calculateNumberSize,
  calculateStringSize,
  concatBytes,
  createBigInt,
  createBytes,
  createDataView,
  createDeferredPromise,
  createHex,
  createModuleLogger,
  createNumber,
  createProjectLogger,
  createSandbox,
  directoryExists,
  ensureDirectoryStructureExists,
  exactOptional,
  fileExists,
  forceRemove,
  getChecksumAddress,
  getErrorMessage,
  getJsonRpcIdValidator,
  getJsonSize,
  getKnownPropertyNames,
  getSafeJson,
  gtRange,
  gtVersion,
  hasProperty,
  hexToBigInt,
  hexToBytes,
  hexToNumber,
  inMilliseconds,
  isASCII,
  isBytes,
  isCaipAccountAddress,
  isCaipAccountId,
  isCaipChainId,
  isCaipNamespace,
  isCaipReference,
  isErrorWithCode,
  isErrorWithMessage,
  isErrorWithStack,
  isHexString,
  isJsonRpcError,
  isJsonRpcFailure,
  isJsonRpcNotification,
  isJsonRpcRequest,
  isJsonRpcResponse,
  isJsonRpcSuccess,
  isNonEmptyArray,
  isNullOrUndefined,
  isObject,
  isPendingJsonRpcResponse,
  isPlainObject,
  isStrictHexString,
  isValidChecksumAddress,
  isValidHexAddress,
  isValidJson,
  isValidSemVerRange,
  isValidSemVerVersion,
  jsonrpc2,
  numberToBytes,
  numberToHex,
  object,
  parseCaipAccountId,
  parseCaipChainId,
  readFile,
  readJsonFile,
  remove0x,
  satisfiesVersionRange,
  signedBigIntToBytes,
  stringToBytes,
  timeSince,
  valueToBytes,
  wrapError,
  writeFile,
  writeJsonFile
};
