// @metamask/utils/dist/hex.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});













var _chunkQEPVHEP7js = require('./chunk-QEPVHEP7.js');
require('./chunk-6ZDHSOUV.js');
require('./chunk-IZC266HS.js');
require('./chunk-QVEKZRZ2.js');
require('./chunk-3W5G4CYI.js');














exports.HexAddressStruct = _chunkQEPVHEP7js.HexAddressStruct; exports.HexChecksumAddressStruct = _chunkQEPVHEP7js.HexChecksumAddressStruct; exports.HexStruct = _chunkQEPVHEP7js.HexStruct; exports.StrictHexStruct = _chunkQEPVHEP7js.StrictHexStruct; exports.add0x = _chunkQEPVHEP7js.add0x; exports.assertIsHexString = _chunkQEPVHEP7js.assertIsHexString; exports.assertIsStrictHexString = _chunkQEPVHEP7js.assertIsStrictHexString; exports.getChecksumAddress = _chunkQEPVHEP7js.getChecksumAddress; exports.isHexString = _chunkQEPVHEP7js.isHexString; exports.isStrictHexString = _chunkQEPVHEP7js.isStrictHexString; exports.isValidChecksumAddress = _chunkQEPVHEP7js.isValidChecksumAddress; exports.isValidHexAddress = _chunkQEPVHEP7js.isValidHexAddress; exports.remove0x = _chunkQEPVHEP7js.remove0x;
