// @metamask/providers/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkKUKZKOBUjs = require('./chunk-KUKZKOBU.js');



var _chunkIZY7ABOLjs = require('./chunk-IZY7ABOL.js');


var _chunkDD6YP3BVjs = require('./chunk-DD6YP3BV.js');



var _chunkWBB62AKCjs = require('./chunk-WBB62AKC.js');



var _chunkHP7EYLLYjs = require('./chunk-HP7EYLLY.js');
require('./chunk-Q4DN6VYN.js');


var _chunkDWR5HIZKjs = require('./chunk-DWR5HIZK.js');


var _chunkA3W22U42js = require('./chunk-A3W22U42.js');
require('./chunk-O5ECOCX2.js');
require('./chunk-6QNVTE4W.js');
require('./chunk-ZOFGBGOM.js');
require('./chunk-4EQNSGSR.js');
require('./chunk-3W5G4CYI.js');











exports.BaseProvider = _chunkA3W22U42js.BaseProvider; exports.MetaMaskInpageProvider = _chunkHP7EYLLYjs.MetaMaskInpageProvider; exports.MetaMaskInpageProviderStreamName = _chunkHP7EYLLYjs.MetaMaskInpageProviderStreamName; exports.StreamProvider = _chunkDWR5HIZKjs.StreamProvider; exports.createExternalExtensionProvider = _chunkKUKZKOBUjs.createExternalExtensionProvider; exports.eip6963AnnounceProvider = _chunkWBB62AKCjs.announceProvider; exports.eip6963RequestProvider = _chunkWBB62AKCjs.requestProvider; exports.initializeProvider = _chunkIZY7ABOLjs.initializeProvider; exports.setGlobalProvider = _chunkIZY7ABOLjs.setGlobalProvider; exports.shimWeb3 = _chunkDD6YP3BVjs.shimWeb3;
